#!/usr/bin/env python3

def test1():   # Объявление функции без параметров
    print(a)   # Обращение к глобальной переменной a


def test2():   # Объявление функции без параметров
    a = 2      # Объявлена локальная переменная a, имеющая такое же имя, как глобальная переменная
    print(a)   # Обращение к локальной переменной a
    del a      # Удаление локальной переменной


def test3(a):  # Объявление функции с параметром a, имеющим такое же имя, как глобальная переменная a
    a = a + 1  # Обращение к локальной переменной a
    print(a)   # Обращение к локальной переменной a
    del a      # Удаление локальной переменной b


def test4():   # Объявление функции без параметров
    print(a)   # Обращение к глобальной переменной a
    # del a    # line 20, in test4 UnboundLocalError: local variable 'a' referenced before assignment


def test5():   # Объявление функции без параметров
    b = a      # Объявлена локальная переменная b, имеющим такое же имя, как глобальная переменная b
    print(b)   # Обращение к локальной переменной b
    del b      # Удаление локальной переменной b


def test6(a):  # Объявление функции с параметром a, имеющим такое же имя, как глобальная переменная a
    b = a      # Объявлена локальная переменная b, имеющая такое же имя, как глобальная переменная b
    print(b)   # Обращение к локальной переменной b
    del b      # Удаление локальной переменной b


# Начало программы
a = 1          # Глобальная переменная a
b = 2          # Глобальная переменная b

test1()        # Выведет: 1
test2()        # Выведет: 2
test3(a)       # Выведет: 2
test4()        # Выведет: 1
test5()        # Выведет: 1
test6(b)       # Выведет: 2
print(b)       # Выведет: 2
